cdk/constructs/DpsStacItemGenerator/runtime/src/dps_stac_item_generator/item.py

```diff
@@ -0,0 +1,65 @@
+import logging
+from typing import Any, Generator, Union
+from urllib.parse import urlparse
+
+import boto3
+import pystac
+from pystac import Link
+from pystac.stac_io import DefaultStacIO, StacIO
+from stac_pydantic.item import Item
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+class CustomStacIO(DefaultStacIO):
+    def __init__(self):
+        self.s3 = boto3.resource("s3")
+        super().__init__()
+
+    def read_text(self, source: Union[str, Link], *args: Any, **kwargs: Any) -> str:
+        parsed = urlparse(str(source))
+        if parsed.scheme == "s3":
+            bucket = parsed.netloc
+            key = parsed.path[1:]
+
+            obj = self.s3.Object(bucket, key)
+            return obj.get()["Body"].read().decode("utf-8")
+        else:
+            return super().read_text(source, *args, **kwargs)
+
+    def write_text(
+        self, dest: Union[str, Link], txt: str, *args: Any, **kwargs: Any
+    ) -> None:
+        parsed = urlparse(str(dest))
+        if parsed.scheme == "s3":
+            bucket = parsed.netloc
+            key = parsed.path[1:]
+            self.s3.Object(bucket, key).put(Body=txt, ContentEncoding="utf-8")
+        else:
+            super().write_text(dest, txt, *args, **kwargs)
+
+
+StacIO.set_default(CustomStacIO)
+
+
+def get_stac_items(catalog_json_key: str) -> Generator[Item, Any, Any]:
+    """
+    Yield STAC items out of a catalog.json
+    """
+    parsed = urlparse(catalog_json_key)
+
+    # get username out of s3 key
+    collection_id = parsed.path.split("/")[1]
+
+    catalog = pystac.Catalog.from_file(catalog_json_key)
```

Comment:
does pystac handle validating the input, or was the try catch outside of this context?

Reply:
If this function (`get_stac_items`) raises any errors when reading the catalog.json or parsing/validating the items, the item generation process will break and will send the failed input message (catalog.json S3 key) to a Dead Letter Queue after a few retries.

I added a few tests to ensure that the right exceptions are raised by the function in the case of bad inputs.